src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,382 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
 // =======================================================================================
 // TextEncoder implementation
 
+namespace {
+
+constexpr inline bool isLeadSurrogate(char16_t c) {
+  return 0xD800 <= c && c < 0xDC00;
+}
+
+constexpr inline bool isTrailSurrogate(char16_t c) {
+  return 0xDC00 <= c && c <= 0xDFFF;
+}
+
+// Calculate UTF-8 length from UTF-16 with potentially invalid surrogates.
+// Invalid surrogates are counted as U+FFFD (3 bytes in UTF-8).
+// Uses SIMD for valid portions and falls back to scalar for invalid surrogates.
+size_t utf8LengthFromInvalidUtf16(kj::ArrayPtr<const char16_t> input) {
+  size_t inputPos = 0;
+  size_t utf8Length = 0;
+
+  while (inputPos < input.size()) {
+    // Find the next invalid surrogate using SIMD validation
+    auto result =
+        simdutf::validate_utf16_with_errors(input.begin() + inputPos, input.size() - inputPos);
+
+    if (result.error == simdutf::error_code::SUCCESS) {
+      // Remaining input is valid - calculate length with SIMD
+      utf8Length +=
+          simdutf::utf8_length_from_utf16(input.begin() + inputPos, input.size() - inputPos);
+      break;
+    }
+
+    if (result.error == simdutf::error_code::SURROGATE) {
+      // Calculate length for the valid portion before the error with SIMD
+      if (result.count > 0) {
+        utf8Length += simdutf::utf8_length_from_utf16(input.begin() + inputPos, result.count);
+        inputPos += result.count;
+      }
+
+      // Handle the invalid surrogate at inputPos
+      char16_t c = input[inputPos];
+      if (isLeadSurrogate(c) && inputPos + 1 < input.size() &&
+          isTrailSurrogate(input[inputPos + 1])) {
```

Comment:
Hmm.. if `result.error == simdutf::error_code::SURROGATE` ... is this branch actually possible? (same as the other instance I already commented on)